### PR TITLE
FiLM-Conditioned Fore-Foil SRF: Shape-Aware Correction for NACA6416 Gap

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The aft-foil Surface Refinement (SRF) head delivered -3.0% p_tan by applying a correction MLP conditioned on gap/stagger geometry. The fore-foil surface has a complementary problem: NACA6416 is the primary OOD geometry (as fore-foil in tandem configs), and it has a distinct DSDF1 gradient profile that the model has not been explicitly conditioned to handle.

Two previous fore-foil SRF attempts (#2117, #2124) both **worsened** p_tan because they applied **unconditional** corrections — a fixed learned per-position bias that overfits training shapes and hurts NACA6416 OOD transfer.

**Key differentiator:** Condition the fore-foil correction MLP on a compact DSDF1 shape fingerprint (mean + std of DSDF1 magnitude over fore-foil surface nodes) concatenated with gap/stagger. This 4D conditioning signal distinguishes NACA6416 (distinct DSDF1 stats) from other training foil shapes, allowing shape-specific corrections instead of fixed bias. The mechanism mirrors how the aft-foil SRF head succeeded (conditioning on geometry context unlocks the gain that unconditioned correction cannot).

**Expected impact:** p_tan -2 to -5% (i.e., 28.60 → 27.2–28.0).

### Overview
Add a new `ForeAftFoilRefinementHead` class (mirroring `AftFoilRefinementHead`) that applies an additive correction to fore-foil surface nodes in tandem samples, conditioned on a 4D FiLM signal: `[dsdf1_mean, dsdf1_std, gap, stagger]`.

### Step 1: Define the ForeAftFoilRefinementHead class

Near the `AftFoilRefinementHead` class definition (lines ~440–520 in train.py), add a new class:

```python
class ForeAftFoilRefinementHead(nn.Module):
    """
    FiLM-conditioned fore-foil surface correction MLP.
    Conditioning: [dsdf1_mean, dsdf1_std, gap, stagger] (4D).
    Mirrors AftFoilRefinementHead with 4D instead of 2D conditioning.
    """
    def __init__(self, hidden_dim, pred_dim, film_hidden=32, mlp_hidden=192, mlp_layers=3):
        super().__init__()
        # FiLM network: 4D cond -> scale/shift for hidden_dim
        self.film_mlp = nn.Sequential(
            nn.Linear(4, film_hidden), nn.SiLU(),
            nn.Linear(film_hidden, 2 * hidden_dim)
        )
        # Main correction MLP
        layers = [nn.Linear(hidden_dim, mlp_hidden), nn.SiLU()]
        for _ in range(mlp_layers - 2):
            layers += [nn.Linear(mlp_hidden, mlp_hidden), nn.SiLU()]
        layers += [nn.Linear(mlp_hidden, pred_dim)]
        self.mlp = nn.Sequential(*layers)
        # Zero-init output to start as identity correction
        nn.init.zeros_(self.mlp[-1].weight)
        nn.init.zeros_(self.mlp[-1].bias)

    def forward(self, hidden, pred, cond):
        # hidden: [M, hidden_dim], pred: [M, pred_dim], cond: [M, 4]
        film = self.film_mlp(cond)
        scale, shift = film.chunk(2, dim=-1)
        h = hidden * (1 + scale) + shift
        return self.mlp(h)
```

### Step 2: Add config flags

In the `Config` dataclass, add:
```python
fore_foil_srf_film: bool = False
fore_foil_srf_hidden: int = 192
fore_foil_srf_layers: int = 3
```

And in the argparse block, add:
```python
parser.add_argument('--fore_foil_srf_film', action='store_true')
parser.add_argument('--fore_foil_srf_hidden', type=int, default=192)
parser.add_argument('--fore_foil_srf_layers', type=int, default=3)
```

### Step 3: Instantiate the head

In the model/head instantiation block (near where `aft_srf_head` is created), add:
```python
fore_srf_head = None
if cfg.fore_foil_srf_film:
    pred_dim = 3  # p_in, p_tan, p_re (same as AftFoilRefinementHead)
    fore_srf_head = ForeAftFoilRefinementHead(
        hidden_dim=model.hidden_dim,  # same as aft head
        pred_dim=pred_dim,
        film_hidden=32,
        mlp_hidden=cfg.fore_foil_srf_hidden,
        mlp_layers=cfg.fore_foil_srf_layers
    ).to(device)
    # Compile if using torch.compile
    if cfg.use_compile:
        fore_srf_head = torch.compile(fore_srf_head)
```

Add fore_srf_head parameters to the optimizer (same lr as aft_srf_head):
```python
if fore_srf_head is not None:
    _fore_params = list(fore_srf_head.parameters())
    base_opt.add_param_group({'params': _fore_params, 'lr': _base_lr})
```

### Step 4: Compute fore-foil mask and FiLM conditioning

In the training loop, after `_aft_foil_mask` is computed (and before aft-foil correction is applied), add:
```python
# Fore-foil mask: surface nodes on foil-1 in tandem samples
_fore_foil_mask = is_surface & (_raw_saf_norm <= 0.005) & _is_tandem.unsqueeze(1)

# FiLM conditioning: DSDF1 shape fingerprint + gap/stagger
if fore_srf_head is not None and _fore_foil_mask is not None:
    # DSDF1 magnitude over fore-foil surface nodes
    _dsdf1_mag = x[:, :, 2:6].norm(dim=-1)  # [B, N] — use RAW x (before standardization)
    _fore_mask_f = _fore_foil_mask.float()
    _fore_count = _fore_mask_f.sum(dim=1).clamp(min=1)
    _fore_dsdf1_mean = (_dsdf1_mag * _fore_mask_f).sum(dim=1) / _fore_count  # [B]
    _fore_dsdf1_std = (
        ((_dsdf1_mag - _fore_dsdf1_mean.unsqueeze(1)) ** 2 * _fore_mask_f).sum(dim=1) / _fore_count
    ).sqrt()  # [B]
    _fore_film_cond = torch.stack(
        [_fore_dsdf1_mean, _fore_dsdf1_std, _raw_gap_stagger[:, 0], _raw_gap_stagger[:, 1]], dim=-1
    )  # [B, 4]
```

**Important:** Use raw (pre-standardization) `x[:, :, 2:6]` for the DSDF1 stats — this gives physically meaningful absolute magnitudes.

### Step 5: Apply fore-foil correction

After the aft-foil correction block, add:
```python
if fore_srf_head is not None and model.training and _fore_foil_mask is not None:
    fore_idx = _fore_foil_mask.nonzero(as_tuple=False)  # [M, 2]
    if fore_idx.numel() > 0:
        fore_hidden = hidden[fore_idx[:, 0], fore_idx[:, 1]]  # [M, hidden_dim]
        fore_pred   = pred[fore_idx[:, 0], fore_idx[:, 1]]    # [M, pred_dim]
        _fore_cond  = _fore_film_cond[fore_idx[:, 0]]          # [M, 4]
        with torch.amp.autocast("cuda", dtype=torch.bfloat16):
            fore_correction = fore_srf_head(fore_hidden, fore_pred, _fore_cond).float()
        pred = pred.clone()
        pred[fore_idx[:, 0], fore_idx[:, 1]] += fore_correction
```

### Step 6: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern \
  --wandb_name "fern/fore-srf-film-s42" --wandb_group phase6/fore-srf-film \
  --fore_foil_srf_film --fore_foil_srf_hidden 192 --fore_foil_srf_layers 3 \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, wandb_name "fern/fore-srf-film-s73")
```

### What to report

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| fore_srf_film | 42 | | | | | | |
| fore_srf_film | 73 | | | | | | |
| **2-seed avg** | — | | | | | | |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | — | d7l91p0x, j9btfx09 |

Also report: VRAM usage, whether the fore-foil correction amplitude grows during training (log the correction norm if possible).

**Key diagnostic:** If p_tan WORSENS (like #2117/#2124), check whether the fore-foil mask selects any nodes (log `_fore_foil_mask.sum()`). If it's zero, the mask condition needs adjustment.

## Baseline

| Metric | 2-seed target |
|--------|--------------|
| p_in   | < 13.05      |
| p_oodc | < 7.70       |
| p_tan  | **< 28.60**  |
| p_re   | < 6.55       |

Current best: PR #2130 (GSB + PCGrad). W&B: d7l91p0x (seed 42, p_tan=28.9), j9btfx09 (seed 73, p_tan=28.3).

```bash
cd cfd_tandemfoil && python train.py --agent <name> --wandb_name "<name>/baseline" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT fern:

## Results

### Summary

| Config | Seed | p_in | p_oodc | p_tan | p_re | VRAM | W&B |
|--------|------|------|--------|-------|------|------|-----|
| fore_srf_film | 42 | 13.7 | 7.9 | 29.1 | 6.6 | 45.8GB | zr273q9f |
| fore_srf_film | 73 | 13.9 | 7.7 | 29.5 | 6.6 | 46.6GB | vzxare6h |
| **2-seed avg** | — | **13.80** | **7.80** | **29.30** | **6.60** | — | — |
| **Baseline** | — | 13.05 | 7.70 | 28.60 | 6.55 | — | d7l91p0x, j9btfx09 |

**Negative result.** All metrics regressed vs. baseline: p_in +5.7%, p_tan +2.4%, p_oodc +1.3%, p_re +0.8%.

### Commands used

```bash
# Seed 42
cd cfd_tandemfoil && python train.py --agent fern   --wandb_name "fern/fore-srf-film-s42" --wandb_group phase6/fore-srf-film   --fore_foil_srf_film --fore_foil_srf_hidden 192 --fore_foil_srf_layers 3   --seed 42   --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4   --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96   --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5   --cosine_T_max 160 --pressure_first --pressure_deep   --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3   --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05   --pcgrad_3way --pcgrad_extreme_pct 0.15 --gap_stagger_spatial_bias

# Seed 73 (same, --seed 73, --wandb_name "fern/fore-srf-film-s73")
```

### Training details

- Epochs: 148 (s42), 149 (s73) — wall-clock limit at 180 min
- Head params: 87,523 (fore_srf_film) + 113,475 (aft_srf) + 113,475 (refine) = additional ~315K over main model

### Diagnostic: correction norm collapsed to zero

The key diagnostic is the **fore-foil correction norm**, which starts at a moderate value then trends to near-zero within ~20 epochs (W&B: `train/fore_srf_correction_norm` shows ▁ at end of training for both seeds). This means the FiLM-conditioned head learned to apply near-zero corrections — essentially becoming a no-op — rather than learning useful shape-specific adjustments.

This mirrors what happened in #2117 (split fore-foil) and #2124 (stacked fore-foil): the fore-foil head consistently fails to learn useful residuals. The pattern across three attempts suggests that:

1. **Limited tandem training signal**: Fore-foil (ID=6) nodes appear only in tandem samples (~20% of training data). With the main SRF head already handling all surface nodes and the aft-foil head handling ID=7 nodes, the residual signal for fore-foil nodes specifically is very small.

2. **DSDF1 statistics are not discriminative enough**: The shape fingerprint (mean + std of x[:,2:6] norm over fore-foil nodes) likely has low variance across training foil shapes, so the FiLM conditioning can't distinguish NACA6416 from other geometries and defaults to the zero-correction identity.

3. **Optimizer interference**: Adding a fore-foil head shares optimizer state with the existing heads, and the main model's gradients (dominated by single-foil and volume loss) may push the fore-foil correction toward zero.

### What happened

The FiLM conditioning (4D: dsdf1_mean, dsdf1_std, gap, stagger) provides richer geometry context than the unconditioned approaches in #2117/#2124, but still fails. The correction norm collapse is the decisive indicator — the conditioning signal is insufficient to drive useful learning. This is the third consecutive negative result on fore-foil corrections.

### Suggested follow-ups

1. **Abandon the fore-foil SRF direction**: Three experiments (split, stacked, FiLM-conditioned) have all failed. The fundamental problem is insufficient training signal for fore-foil-specific corrections. The main model + aft-foil SRF may already be near the limit for this architecture.

2. **If fore-foil SRF is pursued further**, the conditioning needs to be richer — e.g., using the full NACA digit encoding (x[:,15:18]) directly as part of the FiLM signal rather than derived statistics, and training with much higher fore-foil surface weight to overcome signal sparsity.

3. **Focus on global improvements**: Given consistent failures on additive corrections, the next improvements might come from architectural changes to the main model (better attention mechanisms for multi-foil interaction) or more data augmentation targeting the tandem distribution.
